client/batch.go

```diff
@@ -77,91 +77,108 @@ func (b *Batch) initResult(calls, numRows int, err error) {
 	b.Results = append(b.Results, r)
 }
 
-func (b *Batch) fillResults() error {
+func (b *Batch) fillResults(br *proto.BatchResponse, pErr *proto.Error) error {
 	offset := 0
 	for i := range b.Results {
 		result := &b.Results[i]
 
 		for k := 0; k < result.calls; k++ {
-			call := b.calls[offset+k]
+			args := b.reqs[offset+k]
 
+			var reply proto.Response
 			if result.Err == nil {
-				result.Err = call.Reply.Header().GoError()
+				result.Err = pErr.GoError()
+				if result.Err == nil {
+					if offset+k < len(br.Responses) {
+						reply = br.Responses[offset+k].GetValue().(proto.Response)
+					} else if args.Method() != proto.EndTransaction {
+						// TODO(tschottdorf): EndTransaction is excepted here
+						// because it may be elided (r/o txns). Might prefer to
+						// simulate an EndTransaction response instead; this
+						// effectively just leaks here.
+						panic("not enough responses for calls")
+					}
+				}
 			}
 
-			switch t := call.Reply.(type) {
-			case *proto.GetResponse:
+			switch req := args.(type) {
+			case *proto.GetRequest:
 				row := &result.Rows[k]
-				row.Key = []byte(call.Args.(*proto.GetRequest).Key)
+				row.Key = []byte(req.Key)
 				if result.Err == nil {
-					row.Value = t.Value
+					row.Value = reply.(*proto.GetResponse).Value
```

Comment:
For now, I'm not too concerned about it (implicit assertion), but it would
make sense to do basic validation at the boundaries to have some guarantees
when production approaches.
On Sep 24, 2015 21:38, "Peter Mattis" <notifications@github.com> wrote:

> In client/batch.go
> <https://github.com/cockroachdb/cockroach/pull/2653#discussion_r40362711>:
>
> >  				if result.Err == nil {
> > -					row.Value = t.Value
> > +					row.Value = reply.(*proto.GetResponse).Value
>
> These non-checked type assertions will panic if somehow the batch
> responses were misaligned with the requests. Is that a concern? Desired?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2653/files#r40362711>.
>